{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import operator\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from itertools import islice\n",
    "\n",
    "# legalCodons can be customized to only analyse on desired codons \n",
    "legalCodons = [\n",
    "        'TTT', 'TTC', 'TTA', 'TTG', 'CTT', \n",
    "        'CTC', 'CTA', 'CTG', 'ATT', 'ATC', \n",
    "        'ATA', 'ATG', 'GTT', 'GTC', 'GTA', \n",
    "        'GTG', 'TAT', 'TAC', 'TAA', 'TAG', \n",
    "        'CAT', 'CAC', 'CAA', 'CAG', 'AAT', \n",
    "        'AAC', 'AAA', 'AAG', 'GAT', 'GAC', \n",
    "        'GAA', 'GAG', 'TCT', 'TCC', 'TCA', \n",
    "        'TCG', 'CCT', 'CCC', 'CCA', 'CCG', \n",
    "        'ACT', 'ACC', 'ACA', 'ACG', 'GCT', \n",
    "        'GCC', 'GCA', 'GCG', 'TGT', 'TGC', \n",
    "        'TGA', 'TGG', 'CGT', 'CGC', 'CGA', \n",
    "        'CGG', 'AGT', 'AGC', 'AGA', 'AGG', \n",
    "        'GGT', 'GGC', 'GGA', 'GGG']\n",
    "\n",
    "emptyCodonDict = dict(zip(legalCodons, [0] * len(legalCodons)))\n",
    "\n",
    "class codonOverview:\n",
    "    '''\n",
    "    A class that provides codon-centric information in regards\n",
    "    to a given gene list.\n",
    "    \n",
    "    The input for this class is is a .fasta file containing only\n",
    "    CODING SEQUENCES (and the different outputs will be incorrect\n",
    "    if any sequence with e.g. UTR or intronic regions are included).\n",
    "    '''\n",
    "    \n",
    "    def __init__(self, fastaFile):\n",
    "        '''The class initializes by parsing the given .fasta file with biopython,\n",
    "        the parsed fasta file is made into a codon object and a dictionary with\n",
    "        additional information'''\n",
    "        self.parsedFasta = SeqIO.parse(fastaFile, 'fasta')\n",
    "        self.codonObject,self.informationDict = self._countCodons()\n",
    "    \n",
    "    def _countCodons(self):\n",
    "        '''Creates and returns a codon object and informationDict from a fasta\n",
    "        file parsed by Bioppython's SeqIO module.\n",
    "        \n",
    "        Codon object is a dictionary of tuples of dictionaries. The outer dictionary\n",
    "        having gene names as keys and the two inner dictionaries having codons as keys\n",
    "        and number of codons and proportion of codons in respect to the length of the\n",
    "        gene respectivly'''\n",
    "        \n",
    "        codonObject = {}\n",
    "        informationDict = {}\n",
    "    \n",
    "        # Useing Biopython to parse .fasta\n",
    "        for gene in self.parsedFasta:\n",
    "        \n",
    "            # For each gene, initialize by saving:\n",
    "            # Codon dict with all counts = 0 and the sequence,\n",
    "            # and initializing the total number of codons in a sequence to 0\n",
    "            tmpNumberCodonDict = emptyCodonDict.copy()\n",
    "            tmpSequence = gene.seq\n",
    "            tmpNCodonsInGene = 0\n",
    "\n",
    "            informationDict[gene.id] = (gene.description,len(gene.seq),tmpSequence)\n",
    "            \n",
    "            # Make sure that the sequence length divides by 3\n",
    "            if len(tmpSequence) % 3 != 0:\n",
    "                continue\n",
    "        \n",
    "            # Iterate over sequence by for each 3 letters in the sequence,\n",
    "            # adding 1 to the value of the corrosponding key in tmpCodonDict,\n",
    "            # while also adding 1 to tmpNCodons (total number of codons in sequence),\n",
    "            # then removing the codon.\n",
    "            # Repeat untill sequence length = 0\n",
    "            while len(tmpSequence) > 0:\n",
    "                tmpCodon = tmpSequence[0:3]\n",
    "                tmpNumberCodonDict[str(tmpCodon)] += 1\n",
    "                tmpSequence = tmpSequence[3:]\n",
    "                tmpNCodonsInGene += 1\n",
    "            \n",
    "            tmpProportionCodonDict = tmpNumberCodonDict.copy()\n",
    "            \n",
    "            for individualCodon in tmpProportionCodonDict.keys():\n",
    "                tmpProportionCodonDict[individualCodon] /= tmpNCodonsInGene\n",
    "            \n",
    "            # Save as self defined CodonObject:\n",
    "            # A dictionary with gene id as key, and a tuple of two codon dictionaries,\n",
    "            # (one with numbers and one with proportions) as values.\n",
    "            \n",
    "            codonObject[gene.id] = (tmpNumberCodonDict, tmpProportionCodonDict)\n",
    "        return(codonObject, informationDict)\n",
    "    \n",
    "    def HighestOccurrenceCodon(self, query, n = 10, proportionInsteadOfValue = False):\n",
    "        '''Takes:\n",
    "        - query, codon of interest (must be a defined legal codon and must be uppercase letters)\n",
    "        - n, top n genes to return\n",
    "        - proportionInsteadOfValue, if True, the top n genes will be for proportions relative to\n",
    "        gene length and not raw count.\n",
    "        \n",
    "        Outputs a list of tuples of gene names and number/proportion of query'''\n",
    "        \n",
    "        tmpDict = {}\n",
    "    \n",
    "        # Make sure that query is a real codon\n",
    "        if query in legalCodons:\n",
    "        \n",
    "            # For each gene in the codonObject,\n",
    "            # add the codon count for the query and the gene name\n",
    "            # to tmpDict as key and value respectivly.\n",
    "            for geneName, codonTuple in self.codonObject.items():\n",
    "                \n",
    "                if proportionInsteadOfValue == True:\n",
    "                    queryCount = codonTuple[1][query]\n",
    "                else:\n",
    "                    queryCount = codonTuple[0][query]\n",
    "                \n",
    "                tmpDict[geneName] = queryCount\n",
    "            \n",
    "            # Reverse sort tmpDict by codon count\n",
    "            #tmpDict = dict(sorted(tmpDict.items(), reverse = True))\n",
    "            resultList = sorted(tmpDict.items(), key=operator.itemgetter(1), reverse = True)\n",
    "        \n",
    "            # Return top n values from tmpDict\n",
    "            #return(list(islice(tmpDict.items(),n)))\n",
    "            return(resultList[:n])\n",
    "    \n",
    "        else:\n",
    "            print('query: \\'' + query + '\\' is not a legal codon')\n",
    "            \n",
    "    def AverageCodons(self, proportionInsteadOfValue = False):\n",
    "        '''Returns a dictionary with either the number or the proportion of every codon relative to genome'''\n",
    "        avgDict = emptyCodonDict.copy()\n",
    "        count = 0\n",
    "        \n",
    "        for geneName, codonTuple in self.codonObject.items():\n",
    "            \n",
    "            if proportionInsteadOfValue == True:\n",
    "                proportionDict = codonTuple[1].copy()\n",
    "            else:\n",
    "                proportionDict = codonTuple[0].copy()\n",
    "            \n",
    "            for codonName, codonProportion in proportionDict.items():\n",
    "                avgDict[codonName] += codonProportion\n",
    "            count += 1\n",
    "            \n",
    "        for codonName, codonProportion in avgDict.items():\n",
    "            \n",
    "            if proportionInsteadOfValue == True:\n",
    "                avgDict[codonName] = round(codonProportion/count,4)\n",
    "            else:\n",
    "                avgDict[codonName] = round(codonProportion/count,1)\n",
    "        \n",
    "        resultList = sorted(avgDict.items(), key=operator.itemgetter(1), reverse = True)\n",
    "        return(resultList)\n",
    "\n",
    "    def PrintSequence(self, query, queryFormat):\n",
    "        '''Takes gene name and outputs either DNA, RNA or protein sequence'''\n",
    "        \n",
    "        legalPro = ['PRO','PROTEIN','AMINO']\n",
    "        \n",
    "        sequence = self.informationDict[query][2]\n",
    "        \n",
    "        if queryFormat.upper() == 'DNA':\n",
    "            return(sequence)\n",
    "        elif queryFormat.upper() == 'RNA':\n",
    "            return(sequence.transcribe())\n",
    "        elif queryFormat.upper() in legalPro:\n",
    "            return(sequence.translate())\n",
    "        else:\n",
    "            print('ERROR: '+queryFormat.upper()+' is not a valid format')\n",
    "        \n",
    "    def ExportList(self, listOfTuples, outName, outFormat):\n",
    "        '''Takes a list of tuples with gene name and number/proportion\n",
    "        and outputs a .csv file containing ID, number/proportion, gene description and gene length'''\n",
    "        nameList = []\n",
    "        valueList = []\n",
    "        descriptionList = []\n",
    "        lengthList = []\n",
    "                \n",
    "        for i in range(0,len(listOfTuples)):\n",
    "            \n",
    "            name = listOfTuples[i][0]\n",
    "            value = listOfTuples[i][1]\n",
    "            \n",
    "            nameList.append(name)\n",
    "            valueList.append(value)\n",
    "            descriptionList.append(self.informationDict[name][0])\n",
    "            lengthList.append(self.informationDict[name][1])\n",
    "            \n",
    "            if isinstance(value, float) == True:\n",
    "                valueFormat = 'Proportion of codon'\n",
    "            else:\n",
    "                valueFormat = 'Number of codon'\n",
    "            \n",
    "        tmpDict = {'ID': nameList,\n",
    "                   valueFormat: valueList,\n",
    "                   'Description': descriptionList,\n",
    "                   'Gene length': lengthList}\n",
    "\n",
    "        df = pd.DataFrame(tmpDict)\n",
    "        \n",
    "        if outFormat == 'csv':\n",
    "            df.to_csv(outName+'.csv', index = False)\n",
    "        elif outFormat == 'excel':\n",
    "            df.to_excel(outName+'.xlsx', index = False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
